### ISSUE TITLE
Document load order requirements (when anchors.add() should be called)

### ISSUE THREAD

**Issue by k-funk** [bryanbraun/anchorjs#69]
Took me a while to figure out the usage, but in my webpack-built application, I included anchor-js like this:

```
import anchorJS from 'anchor-js';

/* view wrapper */

    var anchors = new anchorJS();
    anchors.options.placement = 'left';
    anchors.add('article h2, article h4');

    // OR
    
    var anchors = new anchorJS({placement: 'left'}).add('article h2, article h4');

/* end view wrapper */

I think it would be helpful to add this example to the README, or if it's incorrect, an example of how you think it should be used correctly.

**Comment by bryanbraun** [comment]
@k-funk, this is a great idea. We forgot to include more usage examples when we added support for modules (webpack, etc). I think the best place for something like this is on the `gh-pages` branch, under the "Basic Usage" section, if we can keep it concise, or "Advanced Usage" if we want a full example with options, like yours above.

**Comment by k-funk** [comment]
@bryanbraun Turns out that my example doesn't actually implement all of anchorjs functionality. Linking someone to anchors doesn't work since the id isn't on the page at load time. I actually don't understand how AnchorJS does this in the first place. If the fragment isn't part of the initial html, how does AnchorJS navigate to an ID that it created, after the page is done loading? I don't see any of that in your source files.

**Comment by bryanbraun** [comment]
@k-funk, the IDs don't need to be in the original HTML payload. Browsers need to complete a number of actions in the loading lifecycle before they trigger the "jump" behavior based on the URL fragment. For example, they can't jump before the DOM is fully loaded, otherwise they might not find the ID. Also, if they jump too early in the rendering process, they might jump to the wrong place on the page.

All of this gives you time to run JS and add the IDs into the markup.

Different browsers jump at different times, but the recommended way to include AnchorJS is before the closing body tag of your HTML document. This means it'll run early in the process, before DOMContentLoaded is complete.

I've done some rough testing to see "how late" I can run AnchorJS in various browsers and have it "still work". In case you are curious, here's what I found (the browser name is listed where they last successfully jumped to the anchor):

|  | on DOMContentLoaded | on window.onload | on arbitrary setTimeout |
| ------------- | ------------- | ------------- | ------------- |
| Chrome  | ✓ | ✘ | ✘ |
| Firefox  | ✓ | ✓ | ✘ |
| Safari  | ✓ | ✘ | ✘ |
| IE  | ✓ | ✓ | ✘ |
| Opera  | ✓ | ✘ | ✘ |

In short, you're good if you run AnchorJS before or on DOMContentLoaded.

The last thing I'll say, is that there's a number of people who use AnchorJS _without_ adding IDs to the page. They often have tools that generate IDs on the backend but they use AnchorJS as an easy way to sprinkle the UI links onto the page. That's a valid use-case that ignores the load-time situation altogether.

**Comment by k-funk** [comment]
I've [created a sample repo](https://github.com/k-funk/simple-webpack-react-starter) using webpack/jquery/anchorjs for you to see my behavior.

From what I'm gathering, using jquery's

    var anchorJS = require('anchor-js');
    var $ = require('jquery');

    $(function() {
      new anchorJS({placement: 'right'}).add('h3');
    });

is going to be too late?

A lot of web applications start after that event.

If it's required to use

    window.addEventListener('DOMContentLoaded', function(){
      new anchorJS({placement: 'right'}).add('h3');
    });

then it seems like that should be well documented, since it's very uncommon to listen to that particular event, in my experience.

**Comment by bryanbraun** [comment]
@k-funk, I haven't any experienced issues testing it in a `DOMContentLoaded` callback, so that's why it hasn't been documented as a requirement thus far.

Thanks for putting together that sample repo. I'm in the middle of a busy couple of weeks right now (ran a marathon yesterday, going to a wedding in a couple of days), but I'll test it out the next time I get a chance.

**Comment by k-funk** [comment]
Correct me if I'm wrong, but it seems like the sam issue is happening in your jsfiddle: https://fiddle.jshell.net/bryanbraun/nc6rL9hk/show/light/#return-of-the-jedi
If you view the source of that page, you'll see that it's inside a `window.onload=function(){ anchors.add('h2'); }`, and that upon refreshing the page with the hash in the url, it doesn't jump to the expected header.

**Comment by bryanbraun** [comment]
@k-funk, ok, I checked out your repo and confirmed that it's not working as expected (and yep, I'm seeing it in that JSFiddle as well). Thanks for putting those together. You're right, we're going to need to add some documentation for this. I've updated the issue title to reflect that (since @afeld put in a PR for the node/commonJS documentation).

I'd like to dig into the details a bit more so we can document it right. A few questions I have are:

- Is there any concern with triggering this inside jQuery's `$(function() { ... }`? I was under the impression that it was the same as `window.addEventListener('DOMContentLoaded` under the hood.
- When in the event cycle is `anchors.add` being triggered in your webpack/react demo app? If after `window.onload`, is that a common usecase?

I'll see what I can find.

We should probably get a better demo than that jsfiddle too... 😆  😞 .

**Comment by bryanbraun** [comment]
I'm doing a more accurate round of testing (and adding the test files to the repo). Here's what I'm finding (updates in progress):

|  | before closing body tag | on DOMContentLoaded | on $(document) .ready |on window .onload | on arbitrary setTimeout |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| Chrome  | | ✓ | | ✘ | ✘ |
| Firefox  | | ✓ | | ✓ | ✘ |
| Safari  | | ✓ | | ✘ | ✘ |
| IE  | | ✓ | ✓ | | ✘ |
| Opera  | | ✓ | | ✘ | ✘ |

**Comment by fulldecent** [comment]
I am not sure if the strong advice on the [documentation](https://www.bryanbraun.com/anchorjs/#overview) should be absolute for everyone.

---

I am [building a site with github pages](https://github.com/fulldecent/github-pages-template). And it uses GitHub Pages' Jekyll, which is kramdown for rendering Markdown --> HTML. This means our headings already automatically get `id="..."` set for them.

So this means we do NOT need anchor.js to add the `id` for us, and we do NOT need anchor.js to jump down the page for us.

---

In our scenario, which is the default for everybody using a statically built Markdown site on GitHub Pages, this means that using the blocking approach in the documentation will add a dependency to painting the page. And this causes us to fail PageSpeed Insights testing.

---

OTOH, if we run anchor.js as deferred, and we enable it in `<script>` after page ready, then there is zero impact on performance.

**Comment by fulldecent** [comment]
So specifically, the warning that including it on `$(document).ready` or later, could result in the links not jumping properly, should be caveated that this only applies if your page does not already have `id="..."` tagging.

**Comment by bryanbraun** [comment]
@fulldecent, oh interesting. You're right, if you already have the IDs then running anchor.js later should be totally fine.

I support adjusting the language in the docs to clarify this.

**Comment by fulldecent** [comment]
Thank you, the docs look great now for this point
